framework/src/transfers/MultiAppGeneralFieldTransfer.C

```diff
@@ -0,0 +1,1209 @@
+
+        if (!_to_boundaries.empty() && !onBoundaries(_to_boundaries, to_moose_mesh, elem))
+          continue;
+
+        // Cache point information
+        // We will use this information later for setting values back to solution vectors
+        cacheOutgoingPointInfo(
+            elem->vertex_average() + _to_positions[i_to], elem->id(), i_to, outgoing_points);
+      } // for
+    }   // else
+  }     // for
+}
+
+void
+MultiAppGeneralFieldTransfer::extractLocalFromBoundingBoxes(std::vector<BoundingBox> & local_bboxes)
+{
+  local_bboxes.resize(_froms_per_proc[processor_id()]);
+  // Find the index to the first of this processor's local bounding boxes.
+  unsigned int local_start = 0;
+  for (processor_id_type i_proc = 0; i_proc < n_processors() && i_proc != processor_id(); ++i_proc)
+    local_start += _froms_per_proc[i_proc];
+
+  // Extract the local bounding boxes.
+  for (unsigned int i_from = 0; i_from < _froms_per_proc[processor_id()]; ++i_from)
+    local_bboxes[i_from] = _bboxes[local_start + i_from];
+}
+
+void
+MultiAppGeneralFieldTransfer::cacheIncomingInterpVals(
+    processor_id_type pid,
+    const unsigned int var_index,
+    std::vector<PointInfo> & pointInfoVec,
+    const std::vector<Point> & point_requests,
+    const std::vector<std::pair<Real, Real>> & incoming_vals,
+    DofobjectToInterpValVec & dofobject_to_valsvec,
+    InterpCaches & interp_caches,
+    InterpCaches & distance_caches)
+{
+  mooseAssert(pointInfoVec.size() == incoming_vals.size(),
+              "Number of dof objects does not equal to the number of incoming values");
+
+  // Get the variable name, with the accomodation for array/vector names
+  const auto & var_name = getToVarName(var_index);
+
+  dof_id_type val_offset = 0;
+  for (auto & pointinfo : pointInfoVec)
+  {
+    // Retrieve target information from cached point infos
+    const auto problem_id = pointinfo.problem_id;
+    const auto dof_object_id = pointinfo.dof_object_id;
+
+    // libMesh EquationSystems
+    auto & es = getEquationSystem(*_to_problems[problem_id], _displaced_target_mesh);
+    // libMesh system
+    System * to_sys = find_sys(es, var_name);
+
+    auto var_num = to_sys->variable_number(var_name);
+    auto & fe_type = to_sys->variable_type(var_num);
+    bool is_nodal = fe_type.family == LAGRANGE;
+
+    // In the higher order elemental variable case, we receive point values, not nodal or elemental
+    // We use an InterpCache to store the values
+    // The distance_cache is necessary to choose between multiple origin problems sending values
+    // This code could be unified with the lower order order case by using the dofobject_to_valsvec
+    if (fe_type.order > CONSTANT && !is_nodal)
+    {
+      // Defining only boundary values will not be enough to describe the variable, disallow it
+      if (_to_boundaries.size())
+        mooseError("Higher order elemental variables are not supported for target-boundary "
+                   "restricted transfers");
```

Comment:
This is a good error for discontinuous (including side-discontinuous) FE types.  But we're not testing for discontinuous types up above, right?  We're testing for non-Lagrange types.  Hierarchic, Bernstein, Szabab, and Rational Bernstein ought to still work fine here, shouldn't they?  Vector types and C1 types I'm less sure about, but if they're broken it's probably not by the boundary restrictions.

Reply:
can any of these families be described in a 3D element by values only on one of its 2D face?
This was my second concern when I banned it

Reply:
Even with a Lagrange, the values on one face don't define the values on the interior.  But the way libMesh does support is the same for all elements; for C0 elements the values on the face's node's DoFs are both necessary and sufficient to define the values on the whole face, but insufficient to define values anywhere except the face.  Hierarchic, Bernstein, and Szabab are the same function space as Lagrange, just with a different basis for p>1, and Rational Bernstein can have different shape functions but is still homeomorphic to the other C0 spaces.

Reply:
Ok I agree it would be fair to support these variable types for target-boundary restriction.

Reply:
Enabled it.